incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSubscribeFactory.java

```diff
@@ -282,52 +346,95 @@ private void onMqttFlush(
 
-            replySeq = sequence;
+            initialSeq = sequence;
 
-            assert replyAck <= replySeq;
+            assert initialAck <= initialSeq;
 
             final OctetsFW extension = flush.extension();
             final MqttFlushExFW mqttFlushEx = extension.get(mqttFlushExRO::tryWrap);
 
             assert mqttFlushEx.kind() == MqttFlushExFW.KIND_SUBSCRIBE;
             final MqttSubscribeFlushExFW mqttSubscribeFlushEx = mqttFlushEx.subscribe();
 
-            filters = mqttSubscribeFlushEx.filters();
-            subscriptionIds.clear();
+            currentFilters = mqttSubscribeFlushEx.filters();
+            messagesSubscriptionIds.clear();
+
+            final KafkaFlushExFW kafkaFlushEx = createKafkaFlushEx(currentFilters);
 
+            messages.doKafkaFlush(traceId, authorization, budgetId, reserved, kafkaFlushEx);
 
-            final KafkaFlushExFW kafkaFlushEx =
-                kafkaFlushExRW.wrap(writeBuffer, FlushFW.FIELD_OFFSET_EXTENSION, writeBuffer.capacity())
-                    .typeId(kafkaTypeId)
-                    .merged(m ->
+            if (retainAvailable)
+            {
+                retainedFilters.removeIf(rf -> !currentFilters.anyMatch(f -> f.pattern().asString().equals(rf.filter)));
+                final List<Subscription> newRetainedFilters = new ArrayList<>();
+                currentFilters.forEach(filter ->
+                {
+                    final boolean sendRetained = (filter.flags() & SEND_RETAIN_FLAG) != 0;
+                    final boolean alreadySentRetained =
+                        retainedFilters.stream().anyMatch(rf -> rf.filter.equals(filter.pattern().asString()));
+                    if (sendRetained && !alreadySentRetained)
+                    {
+                        newRetainedFilters.add(new Subscription(
+                            (int) filter.subscriptionId(), filter.pattern().asString(), filter.qos(), filter.flags()));
+                        final boolean rap = (filter.flags() & RETAIN_AS_PUBLISHED_FLAG) != 0;
+                        retainAsPublished.put((int) filter.subscriptionId(), rap);
+                    }
+                });
+
+                if (!newRetainedFilters.isEmpty())
+                {
+                    if (MqttKafkaState.initialOpened(retained.state) && !MqttKafkaState.initialClosed(retained.state))
+                    {
+                        deferredRetainedFiltersQueue.add(newRetainedFilters);
+                    }
+                    else
                     {
-                        m.capabilities(c -> c.set(KafkaCapabilities.FETCH_ONLY));
-                        filters.forEach(filter ->
+                        retained.state = 0;
+                        retained.replySeq = 0;
+                        retained.replyAck = 0;
+                        retained.replyMax = 0;
+                        retained.doKafkaBegin(traceId, authorization, 0, newRetainedFilters);
```

Comment:
It might be clearer to reinitialize the retained state inside `doKafkaBegin` of `retained` instead of modifying it outside.